File: app/src/am_sangntn/com/example/asian/UserAdapter.java
package com.example.asian;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

public class UserAdapter extends RecyclerView.Adapter<UserAdapter.UserViewHolder> {

    private Context mContext;
    private List<User> mListUser;

    @SuppressLint("NotifyDataSetChanged")
    public void setData(List<User> mListUser) {
        this.mListUser = mListUser;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public UserViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_user, parent, false);
        mContext = parent.getContext();
        return new UserViewHolder(view);
    }

    @SuppressLint("NotifyDataSetChanged")
    @Override
    public void onBindViewHolder(@NonNull UserViewHolder holder, int position) {
        User user = mListUser.get(position);
        if (user == null) {
            return;
        }
        holder.mTvId.setText(String.valueOf(user.getId()));
        holder.mTvAge.setText(String.valueOf(user.getAge()));
        holder.mTvName.setText(user.getName());

        holder.mBtnDelete.setOnClickListener(view -> {
            UserDatabase.getInstance(mContext).userDao().deleteUser(user);
            mListUser.remove(holder.getAdapterPosition());
            notifyDataSetChanged();
        });

    }

    @Override
    public int getItemCount() {
        if (mListUser != null) {
            return mListUser.size();
        }
        return 0;
    }

    public static class UserViewHolder extends RecyclerView.ViewHolder {

        private final TextView mTvName, mTvAge, mTvId;
        private final Button mBtnDelete;

        public UserViewHolder(@NonNull View itemView) {
            super(itemView);

            mTvName = itemView.findViewById(R.id.tvUserName);
            mTvAge = itemView.findViewById(R.id.tvUserAge);
            mBtnDelete = itemView.findViewById(R.id.btnDelete);
            mTvId = itemView.findViewById(R.id.tvUserId);
        }
    }
}

File: app/src/main/java/com/example/ex02/view/ex01/UserAdapter.kt
package com.example.ex02.view.ex01

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.lifecycle.MutableLiveData
import androidx.recyclerview.widget.RecyclerView
import com.example.ex02.databinding.ItemEx01UserBinding
import com.example.ex02.model.ex01.User
import com.example.ex02.viewmodel.ex01.UserViewModel

class UserAdapter(private val userViewModel: UserViewModel) : RecyclerView.Adapter<UserAdapter.UserViewHolder>() {

    private var users = MutableLiveData<List<User>>()


    fun setData(list: List<User>) {
        users.postValue(list)
        notifyDataSetChanged()
    }

    class UserViewHolder(private val binding: ItemEx01UserBinding) : RecyclerView.ViewHolder(binding.root) {
        var deleteButton = binding.btnDelete
        fun bind(user: User) {
            binding.tvUserId.text = user.id.toString()
            binding.tvUserName.text = user.name
            binding.tvUserAge.text = user.age.toString()
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): UserViewHolder {
        val binding = ItemEx01UserBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return UserViewHolder(binding)
    }

    override fun getItemCount(): Int {
        return users.value?.size ?: 0
    }

    override fun onBindViewHolder(holder: UserViewHolder, position: Int) {
        users.value?.let { holder.bind(it[position]) }
        holder.deleteButton.setOnClickListener {
            users.value?.let { it1 -> deleteUser(it1[position]) }
        }
    }

    @SuppressLint("NotifyDataSetChanged")
    private fun deleteUser(user: User) {
        userViewModel.deleteUser(user)
        notifyDataSetChanged()
    }
}